rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py

```diff
@@ -0,0 +1,620 @@
+                                        self.last_known_waypoint_index
+                                        ).location) < 0.5:
+                                    self.on_waypoint =\
+                                        self.last_known_waypoint_index
+                                else:
+                                    self.on_lane = None  # update_off_grid()
+                                    self.on_waypoint = None
+                        # ------------------------ #
+                        # IMPLEMENT YOUR CODE HERE #
+                        # If your robot does not have an API to report the
+                        # remaining travel duration, replace the API call
+                        # below with an estimation
+                        # ------------------------ #
+                        duration = self.api.navigation_remaining_duration()
+                        if self.path_index is not None:
+                            self.next_arrival_estimator(
+                                self.path_index, timedelta(seconds=duration))
+            self.path_finished_callback()
+            self.node.get_logger().info(
+                f"Robot {self.name} has successfully navigated along "
+                f"requested path.")
+
+        self._follow_path_thread = threading.Thread(
+            target=_follow_path)
+        self._follow_path_thread.start()
+
+    def dock(
+            self,
+            dock_name,
+            docking_finished_callback):
+        ''' Docking is very specific to each application. Hence, the user will
+            need to customize this function accordingly. In this example, we
+            assume the dock_name is the same as the name of the waypoints that
+            the robot is trying to dock into. We then call api.start_process()
+            to initiate the robot specific process. This could be to start a
+            cleaning process or load/unload a cart for delivery.
+        '''
+
+        self._quit_dock_event.clear()
+        if self._dock_thread is not None:
+            self._dock_thread.join()
+
+        self.dock_name = dock_name
+        assert docking_finished_callback is not None
+        self.docking_finished_callback = docking_finished_callback
+
+        # Get the waypoint that the robot is trying to dock into
+        dock_waypoint = self.graph.find_waypoint(self.dock_name)
+        assert(dock_waypoint)
+        self.dock_waypoint_index = dock_waypoint.index
+
+        def _dock():
+            # Request the robot to start the relevant process
+            self.node.get_logger().info(
+                f"Requesting robot {self.name} to dock at {self.dock_name}")
+            self.api.start_process(self.dock_name, self.map_name)
+
+            with self._lock:
+                self.on_waypoint = None
+                self.on_lane = None
+            time.sleep(0.5)
+            # ------------------------ #
```

Comment:
So far we have been using docking as a way to initiate cleaning processes. But it doesn't look like that is supported with this fleet adapter. For example, in the `airport_terminal` world, if I request for `zone_1` to be cleaned, once the robot arrives as `zone_1` waypoint, the fleet adapter will call the `dock()` functions. But in the implementation here, it makes a call to `api.start_process()` but the `/open-rmf/rmf_demos_fm/start_task?robot_name={self.robot_name}` endpoint does not trigger anything in `fleet_manager.py` so the robot remains stuck. See related comment in `fleet_manager.py`.

Reply:
We should also subscribe to `/dock_summary` here to get the list of waypoints the robot will navigate through while docking/cleaning. The in the "Implement your code here" block, we should get the [unstable participant](https://github.com/open-rmf/rmf_ros2/blob/46c2bb931342673347adaf0c5b91f2f4ef9441d4/rmf_fleet_adapter_python/src/adapter.cpp#L137-L143) and then [set the itinerary](https://github.com/open-rmf/rmf_ros2/blob/46c2bb931342673347adaf0c5b91f2f4ef9441d4/rmf_fleet_adapter/src/full_control/main.cpp#L439-L443) with the waypoints remaining in the docking/cleaning path.

Reply:
Sorry didn't realise the robot was stuck! Thanks for pointing it out. Added the subscribers in both `fleet_manager` and `RobotCommandHandle`.\
One thing though, `mock_docker` publishes `/dock_summary` before the two subscribers are ready, so the message is often missed. I thought the `qos_profile` specified for this publisher should help with this, but it didn't seem to work for me. I added a `time.sleep(20)` in `mock_docker` before it publishes as a temporary solution, but I'm still looking for a better way in case the manager or command handle takes too long to set up in other cases